rfcs/text/0012_background_sessions.md

```diff
@@ -0,0 +1,291 @@
+
+```ts
+session.getAll()
+```
+
+Retrieves all of the background session saved objects for the current user. (We can filter it to the current user
+because when the saved object is created, we store a hash of the current user's username inside the object.)
+
+---
+
+```ts
+session.get(sessionId)
+```
+
+Retrieves the saved object for this `sessionId` (if there is one), including the request/search ID hash, status
+(complete/incomplete), application data, and restore URL.
+
+---
+
+```ts
+session.save(sessionId, restoreUrl, appData)
+```
+
+Creates the saved object for this background session. Used when a user chooses to "send to background."
+
+---
+
+```ts
+session.delete(sessionId)
+```
+
+Deletes the saved object for this background session, and cancels all incomplete search requests for this session.
+
+---
+
+```ts
+session.getSearchId(sessionId, searchRequest)
+```
+
+Used internally by the search service when a search request is handled to look up the search ID (if there is one) for
+this `sessionId` and `searchRequest`. If bypassing the `search` service, this would need to be called manually when a
+search request is handled.
+
+---
+
+```ts
+session.trackSearchId(sessionId, searchRequest, searchId)
+```
+
+Used internally by the search service when the first response from Elasticsearch is received. After this is called, the
+search service will periodically monitor if a saved object is created for this `sessionId`, and will update it to
+include this `searchRequest`/`searchId` association. If bypassing the `search` service, this would need to be called
+manually when the first response from Elasticsearch is received.
+
+## Public APIs
+
+The following APIs are exposed by the public-side `data` plugin start `search` service:
+
+```ts
+session.start()
+```
+
+Starts a new session and returns the `sessionId`, a unique ID for this session that will be sent in `search` requests.
+
+---
+
+```ts
+session.set(sessionId)
+```
+
+Sets the `sessionId` that is sent in `search` requests to the given ID. Used when restoring a background session so that
+search requests may resume instead of restart.
+
+---
+
+```ts
+session.clear()
+```
+
+Clears the current `sessionId` so that future `search` requests do not contain a `sessionId`. Automatically called when
+users navigate between applications in Kibana.
+
+---
+
+```ts
+session.getAll()
+session.get(sessionId)
+session.save(sessionId, restoreUrl, appData)
+session.delete(sessionId)
+```
+
+These APIS make a request to the corresponding server-side API with the given parameters.
+
+
+# Limitations
+
+In the first iteration, cases which require multiple search requests to be made serially will not be supported. The
+following are examples of such scenarios:
+
+- When a visualization is configured with a terms agg with an "other" bucket
+- When using blended layers or term joins in Maps
+
+Eventually, when expressions can be run on the server, we will facilitate these use cases by adding support for
```

Comment:
i think we shouldn't be adding support for background expressions to this service. Preferably this should be a general solution that can be applied to anything. So i think we should make a section in this RFC describing how server side routes issuing multiple requests can work.

Reply:
This was never the intention, if an expression runs on the server, all of the search requests it's going to execute will be in the context of a single `sessionId`.

Reply:
I fixed the phrasing of this in the RFC.